Introduce tests for crosswords

MatthewJWalls opened the pull request:
<!--
[We'd love some feedback on any struggles you had with this PR](https://goo.gl/forms/QRQaco334CdpfmNF2).
All the questions are optional. Any amount of feedback is great!
-->

## What does this change?

* Separate out the crossword react component to make it accessible directly
* Remove usages of the classNames library and move it inline
    * I was having some problems compiling for test with this dependency, since it was just 5 
       lines of code I inlined it and dropped the dependency.
* Write tests

## What is the value of this and can you measure success?

Now under test

<!--
If setting up an AB test, make sure you have read our [AB testing doc](https://github.com/guardian/frontend/blob/master/docs/03-dev-howtos/01-ab-testing.md).
-->

## Does this affect other platforms - Amp, Apps, etc?

No

<!--
Run the AMP test suite with `make validate-amp`

You should also validate a specific page that your change affects by adding the amp query string along with the development hash: http://localhost:3000/sport/2016/aug/25/katie-ledecky-first-pitch-washington-nationals-bryce-harper?amp=1#development=1

The AMP validation results will appear in your console.
-->

## Screenshots

## Request for comment


<!--
*Does this PR meet the [contributing guidelines](https://github.com/guardian/frontend/blob/issue_pr_templates/.github/CONTRIBUTING.md#submission)?*
-->

dominickendrick commented on static/src/javascripts/projects/common/modules/crosswords/cell.js:
```diff
@@ -16,6 +14,14 @@ define([
 
+        classNames : function(props) {
+            var out = [];
+            for(var f in props){
```
Could this be done with:
```
return Object.keys(props)
 .filter(function(f) {props[f] === true})
 .join(' ')
```
for more functional fun ?

MatthewJWalls replied:
![themoreyouknow](https://media0.giphy.com/media/Y2nbrJyAR6RiM/200_s.gif)

dominickendrick commented on static/src/javascripts/projects/common/modules/crosswords/confirm-button.js:
```diff
@@ -1,20 +1,27 @@
             };
         },
 
+        classNames : function(props) {
```
It does feel a bit odd to have to inline these, could we put the `className` helper in `crossword` module path to see if we can fix the test compilation issues ?

dominickendrick commented:
👍 apart from comments - Great to see more tests !!!

MatthewJWalls commented:
I've cleaned up the classNames dependency and moved it as a package-level module

SiAdcock commented on static/src/javascripts/projects/common/modules/crosswords/classNames.js:
```diff
@@ -0,0 +1,9 @@
+            .join(' ');
+    };
+
+});
```
Elegant 😍 

Since this module has no dependencies, you can drop the empty array from the first argument of the define call.  

```
define(function() { ... })
```

SiAdcock commented on static/src/javascripts/projects/common/modules/crosswords/crossword.js:
```diff
@@ -0,0 +1,720 @@
+    some,
+    zip,
+    forEach,
+    find
```
I wonder how many of these lodash array methods we need any more? Certainly `filter`, `map`, `some` and `forEach` are part of ES5 and supported all the way back to IE9. We include an ES5 shim to support even older browsers.

SiAdcock commented:
Nice, @MatthewJWalls, I'm very excited to see more client side tests! 

I noticed that a number of the tests are testing multiple (related) actions. For instance, `Can move focus with tab keys` tests that `tab` and `shift-tab` actions have the desired result. In my opinion each test should only test one action, so that they can only fail for one reason. If the aforementioned `tab keys` test failed, I wouldn't know if it was the `tab` action or the `shift tab` action that was breaking.

What do you think?

MatthewJWalls commented:
@SiAdcock Yeah I think the same, I'm just fantastically lazy half the time

MatthewJWalls commented:
now done:

* Split tests up @SiAdcock 
* Removed the redundant []
* Fixed a small styling issue I noticed I introduced on the buttons

I think this should do it? I can look at the lodash stuff separate from this, this PR has knocked around for a while now!

SiAdcock commented:
👍

SiAdcock commented:
How about we get this merged @MatthewJWalls? 🚢  🚀

gtrufitt commented:
🙏 We're probably going to have to look at the React stuff soon to refactor out bad habits so would be good to have these!

sndrs commented:
shall we merge this? :

MatthewJWalls commented:
OK I'm just fixing some merge conflicts on this again from the javascript directory change and then I'll force push it again. Fingers crossed everything still works.

PRBuilds commented:
PR build results: 
* https://s3-eu-west-1.amazonaws.com/prbuilds/PR-14956/screenshots/mobile.png 
* https://s3-eu-west-1.amazonaws.com/prbuilds/PR-14956/screenshots/wide.png 
* https://s3-eu-west-1.amazonaws.com/prbuilds/PR-14956/screenshots/tablet.png 
* https://s3-eu-west-1.amazonaws.com/prbuilds/PR-14956/screenshots/desktop.png 

 -automated message

MatthewJWalls commented:
Ok - I'm pretty happy this is good again now. I don't want to change crosswords at the end of the day so I'll merge this tomorrow morning.
